components/hab/src/command/pkg/uninstall/uninstall_impl.rs

```diff
@@ -98,9 +98,16 @@ pub async fn uninstall_all_but_latest<U>(ui: &mut U,
         ui.end(format!("Uninstall of {} complete", ident))?;
         return Ok(0);
     }
-    // sort the idents so the latest occur last in the list
-    idents.sort_unstable_by(|a, b| a.by_parts_cmp(b));
-    let to_uninstall = &mut idents[..(len - number_latest_to_keep)];
+
+    // Reverse sort the idents so the latest occur first in the list. Because we
+    // execute the uninstall hook only when the last revision of a package is
+    // being removed AND we want the most recent uninstall hook to be the one to
+    // run. Performing this sort ensures that the last package to be removed is the
+    // most recent package. This is important because a package's uninstall logic may
+    // include undoing any installation of previous packages.
+    idents.sort_unstable_by(|a, b| b.by_parts_cmp(a));
+
+    let to_uninstall = &mut idents[number_latest_to_keep..];
```

Comment:
I think we can leave the logic that was here before in place, right? I like having the "belt and suspenders" sort in `uninstall_many`, because that really is the code that is most responsible for enforcing this uninstall behavior. This sorting here is still needed to determine what packages we need to deal with, and leaving it as it was means less churn (also, sorting an already sorted list later on in `uninstall_many` should be pretty quick).

This comment would also be best in `uninstall_many` (again, because it is the most responsible for this behavior).

Reply:
sorry @christophermaier  I'm a little confused about what you are asking. This commit does leave the logic that was originally here the same as it was. And it performs a second sort in `uninstall_many`. Are you just asking to move the comment?

Reply:
Ugh, I'm a dork... I was looking at the diff of your commit, and not the overall PR diff. You're right. Yeah, I think just moving the comment is the only thing to do.

Sorry for the confusion! :bow: :man_facepalming:

Reply:
Moved the comment. Yeah that was totally the wrong spot for the comment. I think I left it there by accident.